src/rust/engine/fs/store/src/remote.rs

```diff
@@ -127,94 +201,115 @@ impl ByteStore {
+    let result_future = self.with_byte_stream_client(move |mut client| {
       let resource_name = resource_name.clone();
-      let store = store.clone();
-      async move {
-        let (sender, receiver) = client
-          .write_opt(call_option(&store.headers, None)?.timeout(store.upload_timeout))
-          .map_err(|err| {
-            format!(
-              "Error attempting to connect to upload digest {:?}: {:?}",
-              digest, err
-            )
-          })?;
-
-        let mut sender = sender.sink_compat();
-
-        let chunk_size_bytes = store.chunk_size_bytes;
-        let resource_name = resource_name.clone();
-        let mut stream = futures::stream::unfold((0, false), move |(offset, has_sent_any)| {
-          if offset >= bytes.len() && has_sent_any {
-            futures::future::ready(None)
-          } else {
-            let mut req = bazel_protos::bytestream::WriteRequest::new();
-            req.set_resource_name(resource_name.clone());
-            req.set_write_offset(offset as i64);
-            let next_offset = min(offset + chunk_size_bytes, bytes.len());
-            req.set_finish_write(next_offset == bytes.len());
-            req.set_data(Bytes::from(&bytes[offset..next_offset]));
-            futures::future::ready(Some((
-              Ok((req, grpcio::WriteFlags::default())),
-              (next_offset, true),
-            )))
-          }
-        });
+      let chunk_size_bytes = store.chunk_size_bytes;
 
-        sender.send_all(&mut stream).await.or_else(move |e| {
-          match e {
-            // Some implementations of the remote execution API early-return if the blob has
-            // been concurrently uploaded by another client. In this case, they return a
-            // WriteResponse with a committed_size equal to the digest's entire size before
-            // closing the stream.
-            // Because the server then closes the stream, the client gets an RpcFinished
-            // error in this case. We ignore this, and will later on verify that the
-            // committed_size we received from the server is equal to the expected one. If
-            // these are not equal, the upload will be considered a failure at that point.
-            // Whether this type of response will become part of the official API is up for
-            // discussion: see
-            // https://groups.google.com/d/topic/remote-execution-apis/NXUe3ItCw68/discussion.
-            grpcio::Error::RpcFinished(None) => Ok(()),
-            e => Err(format!(
-              "Error attempting to upload digest {:?}: {:?}",
-              digest, e
-            )),
-          }
-        })?;
+      // TODO(tonic): The call into the Tonic library wants the slice to last for the 'static
+      // lifetime but the slice passed into this method generally points into the shared memory
+      // of the LMDB store which is on the other side of the FFI boundary.
+      let bytes = Bytes::copy_from_slice(bytes);
```

Comment:
Given Tonic needs to pin its future, the byte slice from LMDB needs to be copied since this code is not guaranteed to run under the LMDB transaction.

Is there a way to avoid this copy?

cc @illicitonion @stuhood

Reply:
I wasn't aware of that feature. Will switch to `Bytes`.

Reply:
> But also, below, you're making a `Vec` per uploading chunk... It looks like `prost` supports using `Bytes` instead of `Vec`s for proto fields, which would allow us to sub-slice inside this one `Bytes` copy: https://github.com/danburkert/prost/blob/423f5ec5bd165a7007a388edfb2b485d5bbf40c7/prost-build/src/lib.rs#L261-L316

+1: should do this. It allows for significantly reducing copies off the wire.

Reply:
How would lifetimes work in that case though?

Reply:
The bytes are "on the stack" of the blocked thread the entire time, and the async code that you're blocking on just has a reference to it.

Reply:
Right, but then how does Tonic see a `'static` lifetime. I'm still trying to understand Rust lifetimes. Wouldn't Tonic only see an `'a` lifetime associated with the LMDB transaction which is then imparted to the byte slice as a `&'a bytes` reference?

Reply:
Also the `Bytes` support has not been released yet: https://github.com/danburkert/prost/compare/v0.6.1...master

Reply:
I asked in prost's Discord channel about whether they plan on doing a release.

Reply:
Both things are fine as followups I think, and it'll be easier to inspect once it has landed.

Reply:
Sounds good. Going to resolve this thread, and punt on switching to Bytes for follow-up PR.